[WebXR] Transient input source for immersive AR sessions

djg opened the pull request:
#### 3e2c8c605ad34cd3b580e438ecbc437994e9d411
<pre>
[WebXR] Transient input source for immersive AR sessions
<a href="https://bugs.webkit.org/show_bug.cgi?id=266045">https://bugs.webkit.org/show_bug.cgi?id=266045</a>
<a href="https://rdar.apple.com/119348536">rdar://119348536</a>

Reviewed by Dean Jackson.

Now that we support rendering of immersive AR sessions, we need a way to
interact with them.

This implementation of `transient-pointer`, based upon immersive VR, translates
basic touch interactions into target rays derived from shooting a ray from the
ARCamera position through the touch position in world space.

Introduce a _WKTransientGestureRecognizer that can generate `transient-pointer`
events from UITouch, and attach an instance of that gesture recognizer to the
main view for the immersive AR session.

Access to events is provided via -[WKARPresentationSession collectInputSources],
which provides transient actions suitable for passing to
PlatformXR::Device::FrameData::inputSources

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/ARKitSPI.h: Copied from Source/WebKit/Shared/Cocoa/ARKitSoftLink.h.
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.mm:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
(-[UITouch normalizedLocationInView:]):
(-[_WKARPresentationSession raycastQueryTransformFromPoint:]):
(-[_WKARPresentationSession collectInputSources]):
(-[_WKARPresentationSession gestureRecognizer:shouldReceiveTouch:]):
(-[_WKARPresentationSession loadView]):
(-[_WKTransientAction initWithTargetRay:pose:]):
(-[_WKTransientGestureRecognizer initWithSession:]):
(-[_WKTransientGestureRecognizer touchesBegan:withEvent:]):
(-[_WKTransientGestureRecognizer touchesMoved:withEvent:]):
(-[_WKTransientGestureRecognizer touchesEnded:withEvent:]):
(-[_WKTransientGestureRecognizer touchesCancelled:withEvent:]):
(-[_WKTransientGestureRecognizer _doneWithTouch:]):
(-[_WKTransientGestureRecognizer _platformXRInputSourceFromTransientAction:actionIdentifier:]):
(-[_WKTransientGestureRecognizer collectInputSources]):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e2c8c605ad34cd3b580e438ecbc437994e9d411

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8104 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/30760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30038 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/10240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5395 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/31984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29711 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/25148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/26228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/33324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/7596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3782 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+    if (!_session)
+        return;
+
+    ARFrame* frame = [_session currentFrame];
```
Nit: * on the other side.

cdumez replied:
RetainPtr

djg replied:
From @cdumez comment, I take it we should never have a raw ObjC ptr and let ARC deal with it?

cdumez replied:
If we're using ARC, then raw pointers are fine. If not, then our modern smart pointer rules apply.
I don't know if we're using ARC here. In my experience, ARC is disabled for most of WebKit.

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+
+- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
```
Reminder to remove these when you're ready.

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+
+- (void)_doneWithTouch:(UITouch *)touch
+{
+    dispatch_async(_accessQueue.get(), ^{
```
Would it be worth having this take a list of touches and only dispatch once?

djg replied:
I think so. This is how @achan00 implemented it.

cdumez commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -129,6 +159,21 @@ - (nullable instancetype)initWithSession:(ARSession *)session descriptor:(WKARPr
 
+- (simd_float4x4)raycastQueryTransformFromPoint:(CGPoint)point
+{
+    ARFrame *frame = self.currentFrame;
```
If this isn't using ARC, I think we should be using `RetainPtr` here.

cdumez commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -129,6 +159,21 @@ - (nullable instancetype)initWithSession:(ARSession *)session descriptor:(WKARPr
+    // from the raycast query object.
+    auto allowingTarget = ARRaycastTargetEstimatedPlane;
+    auto alignment = ARRaycastTargetAlignmentAny;
+    ARRaycastQuery *query = [frame raycastQueryFromPoint:point allowingTarget:allowingTarget alignment:alignment];
```
RetainPtr

cdumez commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+    if (!_session)
+        return;
+
+    ARFrame* frame = [_session currentFrame];
```
RetainPtr

cdumez commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+    // space.
+    simd_float4x4 poseTransform = frame.camera.transform;
+
+    UIView* view = [_session view];
```
RetainPtr

cdumez commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
+
+        dispatch_async(_accessQueue.get(), ^{
+            NSNumber *id = @(touch.hash);
+            RetainPtr<_WKTransientAction> transientAction = adoptNS([[_WKTransientAction alloc] initWithTargetRay:targetRayTransform pose:poseTransform]);
```
Doesn't `RetainPtr` work? Maybe it only works with pure C++?

djg committed:
[WebXR] Transient input source for immersive AR sessions
https://bugs.webkit.org/show_bug.cgi?id=266045
rdar://119348536

Reviewed by Dean Jackson.

Now that we support rendering of immersive AR sessions, we need a way to
interact with them.

This implementation of `transient-pointer`, based upon immersive VR, translates
basic touch interactions into target rays derived from shooting a ray from the
ARCamera position through the touch position in world space.

Introduce a _WKTransientGestureRecognizer that can generate `transient-pointer`
events from UITouch, and attach an instance of that gesture recognizer to the
main view for the immersive AR session.

Access to events is provided via -[WKARPresentationSession collectInputSources],
which provides transient actions suitable for passing to
PlatformXR::Device::FrameData::inputSources

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/ARKitSPI.h: Copied from Source/WebKit/Shared/Cocoa/ARKitSoftLink.h.
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.mm:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
(-[UITouch normalizedLocationInView:]):
(-[_WKARPresentationSession raycastQueryTransformFromPoint:]):
(-[_WKARPresentationSession collectInputSources]):
(-[_WKARPresentationSession gestureRecognizer:shouldReceiveTouch:]):
(-[_WKARPresentationSession loadView]):
(-[_WKTransientAction initWithTargetRay:pose:]):
(-[_WKTransientGestureRecognizer initWithSession:]):
(-[_WKTransientGestureRecognizer touchesBegan:withEvent:]):
(-[_WKTransientGestureRecognizer touchesMoved:withEvent:]):
(-[_WKTransientGestureRecognizer touchesEnded:withEvent:]):
(-[_WKTransientGestureRecognizer touchesCancelled:withEvent:]):
(-[_WKTransientGestureRecognizer _doneWithTouch:]):
(-[_WKTransientGestureRecognizer _platformXRInputSourceFromTransientAction:actionIdentifier:]):
(-[_WKTransientGestureRecognizer collectInputSources]):

Canonical link: https://commits.webkit.org/271717@main
